core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesUnassignedShardsAction.java

```diff
@@ -0,0 +1,180 @@
+        final DiscoveryNodes nodes = state.nodes();
+        final MetaData metaData = state.metaData();
+        final RoutingNodes routingNodes = state.routingNodes();
+        Set<String> requestedIndices = new HashSet<>();
+        requestedIndices.addAll(Arrays.asList(request.indices()));
+        List<ShardId> shardIdsToFetch = new ArrayList<>();
+        for (MutableShardRouting shard : Iterables.concat(routingNodes.unassigned(), routingNodes.ignoredUnassigned())) {
+            if (requestedIndices.size() == 0 || requestedIndices.contains(shard.shardId().index().getName())) {
+                shardIdsToFetch.add(shard.shardId());
+            }
+        }
+
+        // async fetch shard status from all the nodes
+        if (shardIdsToFetch.size() > 0) {
+            AsyncShardFetches asyncShardFetches = new AsyncShardFetches(nodes, metaData, shardIdsToFetch.size(), listener);
+            asyncShardFetches.fetch(shardIdsToFetch);
+        } else {
+            listener.onResponse(new IndicesUnassigedShardsResponse());
+        }
+    }
+
+    @Override
+    protected ClusterBlockException checkBlock(IndicesUnassignedShardsRequest request, ClusterState state) {
+        return state.blocks().indicesBlockedException(ClusterBlockLevel.METADATA_READ, request.indices());
+    }
+
+    private class AsyncShardFetches {
+        private final DiscoveryNodes nodes;
+        private final MetaData metaData;
+        private final CountDown expectedOps;
+        private final Queue<InternalAsyncFetch.Response> fetchResponses;
+        private final ActionListener<IndicesUnassigedShardsResponse> listener;
+
+        AsyncShardFetches(DiscoveryNodes nodes, MetaData metaData, int expectedOps, ActionListener<IndicesUnassigedShardsResponse> listener) {
+            this.nodes = nodes;
+            this.metaData = metaData;
+            this.listener = listener;
+            this.expectedOps = new CountDown(expectedOps);
+            this.fetchResponses = new ConcurrentLinkedQueue<>();
+        }
+
+        void fetch(Collection<ShardId> shardIds) {
+            for (ShardId shardId : shardIds) {
+                InternalAsyncFetch fetch = new InternalAsyncFetch(logger, "monitor_shard_copies", shardId, listShardsInfo);
+                fetch.fetchData(nodes, metaData, Collections.<String>emptySet());
+            }
+        }
+
+        private class InternalAsyncFetch extends AsyncShardFetch<TransportNodesListGatewayStartedShards.NodeGatewayStartedShards> {
+
+            InternalAsyncFetch(ESLogger logger, String type, ShardId shardId, TransportNodesListGatewayStartedShards action) {
+                super(logger, type, shardId, action);
+            }
+
+            @Override
+            protected synchronized void processAsyncFetch(ShardId shardId, TransportNodesListGatewayStartedShards.NodeGatewayStartedShards[] responses, FailedNodeException[] failures) {
+                super.processAsyncFetch(shardId, responses, failures);
```

Comment:
I'm doubting about the super call, it just triggers all kind of caching we don't need.